pkg/parse/root_test.go

```diff
@@ -451,6 +454,221 @@ func TestRoot_Parse(t *testing.T) {
 	}
 }
 
+func fakeCRD(opts ...core.MetaMutator) ast.FileObject {
+	//opts = append(opts, core.Name("anvils.acme.com"))
+	crd := fake.CustomResourceDefinitionV1Object(opts...)
+	crd.Spec.Group = "acme.com"
+	crd.Spec.Names = apiextensionsv1.CustomResourceDefinitionNames{
+		Plural:   "anvils",
+		Singular: "anvil",
+		Kind:     "Anvil",
+	}
+	crd.Spec.Scope = apiextensionsv1.ResourceScope(discovery.UnknownScope)
+	crd.Spec.Versions = []apiextensionsv1.CustomResourceDefinitionVersion{
+		{
+			Name:    "v1",
+			Served:  true,
+			Storage: false,
+			Schema: &apiextensionsv1.CustomResourceValidation{
+				OpenAPIV3Schema: &apiextensionsv1.JSONSchemaProps{
+					Type: "object",
+					Properties: map[string]apiextensionsv1.JSONSchemaProps{
+						"spec": {
+							Type:     "object",
+							Required: []string{"lbs"},
+							Properties: map[string]apiextensionsv1.JSONSchemaProps{
+								"lbs": {
+									Type:    "integer",
+									Minimum: pointer.Float64Ptr(1.0),
+									Maximum: pointer.Float64Ptr(9000.0),
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+	return fake.FileObject(crd, "cluster/crd.yaml")
+}
+
+func TestRoot_Parse_Discovery(t *testing.T) {
+	testCases := []struct {
+		name            string
+		format          filesystem.SourceFormat
+		parsed          []ast.FileObject
+		want            []ast.FileObject
+		discoveryClient discovery.ServerResourcer
+		expectError     bool
+	}{
+		{
+			// unknown scoped object does not get skipped when sending go applier
+			name:            "unknown scoped object with discovery failure",
+			format:          filesystem.SourceFormatUnstructured,
+			discoveryClient: syncertest.NewDiscoveryClientWithErrors(true, kinds.Namespace(), kinds.Role()),
+			expectError:     true,
+			parsed: []ast.FileObject{
+				fake.Role(core.Namespace("foo")),
+				// add a faked obect in parser.parsed without CRD so it's scope will be unknown when validating
+				fake.Unstructured(kinds.Anvil(), core.Name("deploy")),
+			},
+			want: []ast.FileObject{},
+		},
+		{
+			// unknown scoped object does not get skipped when sending go applier
+			name:            "unknown scoped object with discovery failure",
+			format:          filesystem.SourceFormatHierarchy,
+			discoveryClient: syncertest.NewDiscoveryClientWithErrors(true, kinds.Namespace(), kinds.Role()),
+			expectError:     true,
```

Comment:
Can you swap these out with an error comparison? That would be better than just a boolean. You can use testutil.AssertEqual with an expected error wrapped with testutil.EqualError to verify both error type and string equivalency, without worrying about exact pointer equivalency.

https://github.com/kubernetes-sigs/cli-utils/blob/master/pkg/testutil/matcher.go#L115

Reply:
Done, using expectedError and doing testutil.AssertEqual with actual parse error.